EA-2453: Add Mongo support for Spring Batch

ikattey opened the pull request:
This PR adds MongoDB support for Spring Batch, and then configures it for the application:

The main changes are in `eu.europeana.entitymanagement.batch.repository`, where the default Spring Batch repository DAOs have been re-implemented with Morphia.

Mongo configuration properties have been changed to the following:

```
mongo.connectionUrl = mongodb://<host>:<port>
mongo.em.database = <entity-management-db>
mongo.batch.database = <spring-batch-repository>
```

ikattey committed:
EA-2453: Add Mongo support for Spring Batch

ikattey committed:
EA-2453: Improvements to Spring Batch support 

- removed dependency on .properties file. Datastore now passed to MongoBatchConfigurer
- removed incorrect maven-resources-plugin config

gsergiu commented on entity-management-batch/pom.xml:
```diff
@@ -0,0 +1,33 @@
+
+    <artifactId>entity-management-batch</artifactId>
+
+    <properties>
```
I think they are defined in the parent pom, we don't need to include them in every module, they should be inherited or access a ${propname} from the parent pom

ikattey replied:
Why do you feel it's better to define this as a property in the parent pom?

gsergiu replied:
it is already property in the parent pom, which was created within the refactoring ticket.... the properties bellow are now duplicates of the parent one, I think they can be safely removed

ikattey replied:
Ah, I see what you mean.

I thought you were referring to the `<artifactId/>`. 

Yes, the properties section can be removed from this file. I'll do that.

gsergiu commented on entity-management-batch/src/main/java/eu/europeana/entitymanagement/batch/BatchConstants.java:
```diff
@@ -0,0 +1,62 @@
+    public static final JobKeyGenerator<JobParameters> JOB_KEY_GENERATOR = new DefaultJobKeyGenerator();
+    public static final InsertManyOptions BATCH_INSERT_OPTIONS = new InsertManyOptions().ordered(true);
+    public static final FindOptions DESCENDING_JOB_EXECUTION = new FindOptions()
+            .sort(descending(JOB_EXECUTION_ID_KEY));
```
looks like LIFO quee, shouldn't we use ascending for FIFO queue?

ikattey replied:
These have nothing to do with queues. 

`BATCH_INSERT_OPTIONS` and `DESCENDING_JOB_EXECUTION` are used to configure Morphia queries.  `JOB_KEY_GENERATOR` is used to generate unique IDs for job instances.

gsergiu replied:
ok I see

gsergiu commented on entity-management-batch/src/main/java/eu/europeana/entitymanagement/batch/config/MongoBatchConfigurer.java:
```diff
@@ -0,0 +1,72 @@
+     * @param datastore Morphia datastore to use
+     */
+    public MongoBatchConfigurer(Datastore datastore) {
+        this.mongoExecutionContextDao = new ExecutionContextRepository(datastore);
```
could we use the datastore as spring bean (with @Resource annotation]?

ikattey replied:
No.

By design, this class needs to be explicitly instantiated. It's also not managed by Spring.

gsergiu replied:
ok

gsergiu commented on entity-management-batch/src/main/java/eu/europeana/entitymanagement/batch/entity/ExecutionContextEntity.java:
```diff
@@ -0,0 +1,75 @@
+        // reproduced from JdbcExecutionContextDao (in Spring batch core)
+        Map<String, Object> map;
+        try {
+            ByteArrayInputStream in = new ByteArrayInputStream(serializedContext.getBytes(StandardCharsets.ISO_8859_1));
```
if I remember correctly StandardCharsets.ISO_8859_1 is Latin1, we have to use UTF8

ikattey replied:
As mentioned in the comment, [this is a direct reproduction from the Spring Batch code](https://github.com/spring-projects/spring-batch/blob/4.3.x/spring-batch-core/src/main/java/org/springframework/batch/core/repository/dao/JdbcExecutionContextDao.java#L322).

Let's trust that they know what they're doing 🤞🏻

gsergiu commented on entity-management-web/pom.xml:
```diff
@@ -38,6 +38,11 @@
 		</dependency>
 
+		<dependency>
+			<groupId>org.springframework.batch</groupId>
```
I assume this is a transitive dependency which will be inherited anyway through entity-management-batch module, or?

ikattey replied:
[Quoting from the Maven docs](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#transitive-dependencies):

> Although transitive dependencies can implicitly include desired dependencies, it is a good practice to explicitly specify the dependencies your source code uses directly.

gsergiu commented on entity-management-web/pom.xml:
```diff
@@ -181,26 +192,6 @@
-							<includes>
-								<include>*.properties</include>
-							</includes>
-							<filtering>true</filtering>
```
resource filtering needs to be activated in order to fill the build.properties with the values from the pom file.  Please revert back this deletion

ikattey replied:
Everything in `src/main/resources` is included in the classpath by default so the filtering is unnecessary. 

It also breaks the application – as everything apart from `.properties` files are excluded from the build. This isn't what we want.

gsergiu commented on entity-management-web/src/main/java/eu/europeana/entitymanagement/EntityManagementApp.java:
```diff
@@ -1,24 +1,31 @@
-        // Remove these exclusions to re-enable security
-        exclude = { SecurityAutoConfiguration.class, ManagementWebSecurityAutoConfiguration.class }
-        )
+        exclude = {
```
do we need these exclusions given the scanBasePackages?

ikattey replied:
These are [AutoConfiguration classes](https://docs.spring.io/spring-boot/docs/1.3.3.RELEASE/reference/html/using-boot-auto-configuration.html).

They are always loaded by default, due to the  `@SpringBootApplication` annotation.

gsergiu commented on entity-management-web/src/main/java/eu/europeana/entitymanagement/batch/DummyJobConfigurer.java:
```diff
@@ -0,0 +1,62 @@
+ * Stub job configurer to test that Spring Batch integration works
+ */
+@Component
+public class DummyJobConfigurer {
```
I suppose that this class is not needed or?

ikattey replied:
Indeed, this is a _temporary_ class which was introduced to test the Spring Batch integration. 

It definitely won't be in the final product.

gsergiu commented on entity-management-web/src/main/java/eu/europeana/entitymanagement/config/AppConfig.java:
```diff
@@ -72,5 +76,15 @@ public ValidatorFactory getValidatorFactoryBean() {
+     * @param datastore Morphia datastore for Spring Batch
+     * @return BatchConfigurer instance
+     */
+    @Bean
```
please assign a bean name

ikattey replied:
In this case, a bean name is unnecessary as:

-  there will only ever be one `BatchConfigurer` within the application
-  this bean is loaded by the Spring context (so we never get to use its qualifier)

gsergiu commented on entity-management-web/src/main/java/eu/europeana/entitymanagement/web/JobLauncherController.java:
```diff
@@ -0,0 +1,46 @@
+public class JobLauncherController {
+
+    @Autowired
+    private DummyJobConfigurer dummyJobConfigurer;
```
if this is really used, I suggest to give the class an appropriate name

ikattey replied:
As mentioned in the class comment, this is a temporary controller used to test the Spring Batch integration.

It will be removed once we start implementing proper Spring Batch jobs.

gsergiu commented on entity-management-web/src/main/java/eu/europeana/entitymanagement/web/JobLauncherController.java:
```diff
@@ -0,0 +1,46 @@
+@RequestMapping(path = "/jobs")
+public class JobLauncherController {
+
+    @Autowired
```
could we use @Resource(name=)

ikattey replied:
See other comment below.

gsergiu commented on entity-management-web/src/main/java/eu/europeana/entitymanagement/web/JobLauncherController.java:
```diff
@@ -0,0 +1,46 @@
+    @Autowired
+    private DummyJobConfigurer dummyJobConfigurer;
+
+    @Autowired
```
could we use @Resource(name=)

ikattey replied:
[The Spring team note that ](https://docs.spring.io/spring-boot/docs/2.0.x/reference/html/using-boot-spring-beans-and-dependency-injection.html)"using `@ComponentScan` (to find your beans) and using `@Autowired` (to do constructor injection) works well."

Constructor-based injection is the recommended way of wiring dependencies.  `@Resource` cannot be used in this manner so it's actually the worse option.

In this case, the better approach would have been to use constructor injection instead.

gsergiu commented on entity-management-web/src/main/java/eu/europeana/entitymanagement/web/JobLauncherController.java:
```diff
@@ -0,0 +1,46 @@
+        JobLauncher jobLauncher = batchConfigurer.getJobLauncher();
+        JobParameters jobParameters = new JobParametersBuilder().addDate(JobParameter.RUN_ID.key(), new Date()).toJobParameters();
+
+        jobLauncher.run(dummyJobConfigurer.helloWorldJob(), jobParameters);
```
was this ment to be an example? Could we move the sample code to a unit test/integration test?

ikattey replied:
This is a temporary controller for testing Spring Batch functionality. 

We could certainly move it to a test, but the purpose here is to provide a quick sanity check that everything works. This controller will be removed in the final product.

gsergiu reviewed (changes requested):
can you please address the comment left behind?

ikattey committed:
EA-2453: Remove duplicate properties in pom.xml

ikattey committed:
EA-2453: Incorporate review remarks

- renamed DummyJobConfigurer to HelloJobConfigurer
- replaced field-based with constructor-based injection
